python/lsst/meas/modelfit/optimizer/optimizer.cc

```diff
@@ -0,0 +1,207 @@
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionInitialSize);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowStepFraction);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionSolverTolerance);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxInnerIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxOuterIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, doSaveIterations);
+    cls.def(py::init<>());
+    return cls;
+}
+
+PyOptimizerHistoryRecorder declareOptimizerHistoryRecorder(py::module & mod) {
+    PyOptimizerHistoryRecorder cls(mod, "OptimizerHistoryRecorder");
+    cls.def(py::init<afw::table::Schema &, std::shared_ptr<Model>, bool>(),
+            "schema"_a, "model"_a, "doRecordDerivatives"_a);
+    cls.def(py::init<afw::table::Schema const &>(), "schema"_a);
+    cls.def("apply", &OptimizerHistoryRecorder::apply,
+            "outerIterCount"_a, "innerIterCount"_a, "history"_a, "optimizer"_a);
+    cls.def(
+        "unpackDerivatives",
+        (void (OptimizerHistoryRecorder::*)(
+            ndarray::Array<Scalar const,1,1> const &,
+            ndarray::Array<Scalar,1,1> const &,
+            ndarray::Array<Scalar,2,2> const &
+        ) const) &OptimizerHistoryRecorder::unpackDerivatives,
+        "nested"_a, "gradient"_a, "hessian"_a
+    );
+    cls.def(
+        "unpackDerivatives",
+        (void (OptimizerHistoryRecorder::*)(
+            afw::table::BaseRecord const &,
+            ndarray::Array<Scalar,1,1> const &,
+            ndarray::Array<Scalar,2,2> const &
+        ) const) &OptimizerHistoryRecorder::unpackDerivatives,
+        "record"_a, "gradient"_a, "hessian"_a
+    );
+    cls.def("fillObjectiveModelGrid", &OptimizerHistoryRecorder::fillObjectiveModelGrid,
+            "record"_a, "parameters"_a, "output"_a);
+    cls.def_readonly("outer", &OptimizerHistoryRecorder::outer);
+    cls.def_readonly("inner", &OptimizerHistoryRecorder::inner);
+    cls.def_readonly("state", &OptimizerHistoryRecorder::state);
+    cls.def_readonly("objective", &OptimizerHistoryRecorder::objective);
+    cls.def_readonly("prior", &OptimizerHistoryRecorder::prior);
+    cls.def_readonly("trust", &OptimizerHistoryRecorder::trust);
+    cls.def_readonly("parameters", &OptimizerHistoryRecorder::parameters);
+    cls.def_readonly("derivatives", &OptimizerHistoryRecorder::derivatives);
+    return cls;
+}
+
+PyOptimizer declareOptimizer(py::module & mod) {
+    PyOptimizer cls(mod, "Optimizer");
+    // StateFlags enum is used as bitflag, so we wrap values as int class attributes.
```

Comment:
I think of using `enum` to define a bitflag in C++ as at best a form of shorthand for defining the same symbols as integer constants, and at worst (in the era of C++11 with `constexpr`) a bit of a hack we should consider moving away from.  Note that with bitflags you don't even get any type safety unless you enumerate all possible combination of bits (which is a bad idea), since you always have to pass the values via some sort of integer.